### PR TITLE
Construct QualifiedFunctionName with CatalogSchemaName and function name

### PR DESCRIPTION
**Pull request by viczhang861**
Do not use dot separated string to create QualifiedFunctionName to lower the chance of throwing runtime error
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by mbasmanova** [comment]
@ajaygeorge 

> is there a difference in the way IllegalArgumentException and PrestoException are handled.

PrestoException with INVALID_FUNCTION_ARGUMENT code can be caught and ignored use `try` function. See https://prestodb.io/docs/current/functions/conditional.html

**Comment by viczhang861** [comment]
> @viczhang861 Vic, could you describe the problem and add a test?

Updated

**Comment by rongrong** [comment]
Where do you see this being a problem? This api should never be called without a check first. So at this level if it doesn't have 3 parts it should be an illegal argument (it's a bug). Currently there doesn't seem to be any call site with less than 3 parts.

**Comment by mbasmanova** [comment]
@viczhang861 Vic, could you share a stack trace from the failed query? @rongrong might have a point that the check is missing somewhere higher up the stack.

File: presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java [hunk 1]
```diff
@@ -38,7 +41,9 @@ public static QualifiedFunctionName of(String dottedName)
     {
         String[] parts = dottedName.split("\\.");
         if (parts.length != 3) {
-            throw new IllegalArgumentException("QualifiedFunctionName should have exactly 3 parts");
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    format("Non-builtin functions must be reference by three parts: catalog.schema.function_name, found: %s", dottedName));
```
Comment (mbasmanova) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: @viczhang861 This class cannot possibly know that it is used to describe a non-builtin function. Can you give an example of a query that is failing?

Comment (viczhang861) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: @mbasmanova 
This is query and stack trace,  https://gist.github.com/viczhang861/77b4fd0e5cafc0ce1e51b3cd6c57febe
After discussion with Rongrong,  check has been added,  add a comment to remind future caller check argument size before using it.

Comment (mbasmanova) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: @viczhang861 

> check has been added, add a comment to remind future caller check argument size before using it.

Is there a PR for that? Perhaps, link it here for future reference. Should we close this PR?

Comment (viczhang861) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: @mbasmanova  #13561 
@caithagoras  I removed reference to this function as you suggested and only allow test code to use it

File: presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java [hunk 2]
```diff
@@ -80,7 +81,11 @@ private SqlInvokedFunction createSqlInvokedFunction(CreateFunction statement)
             throw new PrestoException(GENERIC_USER_ERROR, format("Invalid function name: %s, require exactly 3 parts", statement.getFunctionName()));
         }
 
-        QualifiedFunctionName functionName = QualifiedFunctionName.of(statement.getFunctionName().toString());
+        QualifiedFunctionName functionName = QualifiedFunctionName.of(
+                new CatalogSchemaName(
+                        statement.getFunctionName().getParts().get(0),
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/execution/CreateFunctionTask.java]: nit: for readability, introduce a local variable for `statement.getFunctionName().getParts()` and use it here and in the check above

File: presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java [hunk 3]
```diff
@@ -33,11 +33,13 @@ private QualifiedFunctionName(CatalogSchemaName functionNamespace, String functi
         this.functionName = requireNonNull(functionName, "name is null").toLowerCase(ENGLISH);
     }
 
+    // Do not use this function, it is only used for test or json deserialization
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: Maybe "This function should only be used for JSON deserialization. Do not use it directly."

File: presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java [hunk 4]
```diff
@@ -33,11 +33,13 @@ private QualifiedFunctionName(CatalogSchemaName functionNamespace, String functi
...
     @JsonCreator
     public static QualifiedFunctionName of(String dottedName)
     {
         String[] parts = dottedName.split("\\.");
         if (parts.length != 3) {
+            // this should never happen
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/QualifiedFunctionName.java]: I think `IlligalArgumentException` means that this should never happen. So this is not necessary.

**Comment by rongrong** [comment]
@viczhang861 Do you mind changing the commit title to

`Construct QualifiedFunctionName with CatalogSchemaName and function name`

**Comment by viczhang861** [comment]
> @viczhang861 Do you mind changing the commit title to
> `Construct QualifiedFunctionName with CatalogSchemaName and function name`

Done

**Comment by caithagoras** [comment]
@rongrong Could you help merging this PR?

**Pull request merged**